File: Artemis2MenuBar/TelemetryAPIFetcher.swift
import Foundation

/// Fetches pre-computed telemetry from the Orbital API.
/// Replaces direct AROW + Horizons calls when telemetryEndpoint is configured.
actor TelemetryAPIFetcher {

    struct TelemetryData {
        var altitude: Double = 0
        var distanceToMoon: Double = 0
        var speed: Double = 0
        var isLive: Bool = false
        var phase: String?
        var stale: Bool = false
    }

    private var endpointURL: String

    private let session: URLSession = {
        let config = URLSessionConfiguration.ephemeral
        config.requestCachePolicy = .reloadIgnoringLocalAndRemoteCacheData
        config.urlCache = nil
        return URLSession(configuration: config)
    }()

    init(endpointURL: String) {
        self.endpointURL = endpointURL
    }

    func reconfigure(endpointURL: String) {
        self.endpointURL = endpointURL
    }

    func fetch() async -> TelemetryData? {
        let cacheBust = Int(Date().timeIntervalSince1970)
        let separator = endpointURL.contains("?") ? "&" : "?"
        guard let url = URL(string: "\(endpointURL)\(separator)_=\(cacheBust)") else { return nil }

        do {
            var request = URLRequest(url: url)
            request.timeoutInterval = 15
            request.cachePolicy = .reloadIgnoringLocalAndRemoteCacheData
            request.setValue("no-cache", forHTTPHeaderField: "Cache-Control")
            request.setValue("no-cache", forHTTPHeaderField: "Pragma")

            let (data, response) = try await session.data(for: request)
            guard let http = response as? HTTPURLResponse, http.statusCode == 200 else { return nil }

            return parse(data)
        } catch {
            return nil
        }
    }

    private func parse(_ data: Data) -> TelemetryData? {
        guard let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else { return nil }

        // Reject stale data — fall back to simulation instead of showing outdated values
        if json["stale"] as? Bool == true {
            return nil
        }

        // Also reject if timestamp is more than 5 minutes old
        if let timestamp = json["timestamp"] as? String {
            let df = ISO8601DateFormatter()
            if let date = df.date(from: timestamp),
               Date().timeIntervalSince(date) > 5 * 60 {
                return nil
            }
        }

        var result = TelemetryData()
        result.altitude = json["altitude"] as? Double ?? 0
        result.distanceToMoon = json["distanceToMoon"] as? Double ?? 0
        result.speed = json["speed"] as? Double ?? 0
        result.isLive = json["isLive"] as? Bool ?? false
        result.phase = json["phase"] as? String

        return result
    }
}

File: OrbitalWindows/Services/TelemetryAPIFetcher.cs
using System.Text.Json;

namespace Orbital.Services;

/// <summary>
/// Fetches pre-computed telemetry from the Orbital API.
/// Replaces direct AROW + Horizons calls when telemetryEndpoint is configured.
/// </summary>
public class TelemetryAPIFetcher
{
    public class TelemetryData
    {
        public double Altitude { get; set; }
        public double DistanceToMoon { get; set; }
        public double Speed { get; set; }
        public bool IsLive { get; set; }
        public string? Phase { get; set; }
    }

    private string _endpointUrl;

    private static readonly HttpClient _httpClient = new()
    {
        Timeout = TimeSpan.FromSeconds(15),
        DefaultRequestHeaders =
        {
            { "Cache-Control", "no-cache" },
            { "Pragma", "no-cache" }
        }
    };

    public TelemetryAPIFetcher(string endpointUrl)
    {
        _endpointUrl = endpointUrl;
    }

    public void Reconfigure(string endpointUrl)
    {
        _endpointUrl = endpointUrl;
    }

    public async Task<TelemetryData?> FetchAsync()
    {
        try
        {
            var cacheBust = DateTimeOffset.UtcNow.ToUnixTimeSeconds();
            var separator = _endpointUrl.Contains('?') ? "&" : "?";
            var url = $"{_endpointUrl}{separator}_={cacheBust}";

            var response = await _httpClient.GetAsync(url);
            if (!response.IsSuccessStatusCode)
                return null;

            var json = await response.Content.ReadAsStringAsync();
            return Parse(json);
        }
        catch
        {
            return null;
        }
    }

    private static TelemetryData? Parse(string json)
    {
        try
        {
            using var doc = JsonDocument.Parse(json);
            var root = doc.RootElement;

            return new TelemetryData
            {
                Altitude = root.TryGetProperty("altitude", out var alt) ? alt.GetDouble() : 0,
                DistanceToMoon = root.TryGetProperty("distanceToMoon", out var moon) ? moon.GetDouble() : 0,
                Speed = root.TryGetProperty("speed", out var spd) ? spd.GetDouble() : 0,
                IsLive = root.TryGetProperty("isLive", out var live) && live.GetBoolean(),
                Phase = root.TryGetProperty("phase", out var phase) ? phase.GetString() : null,
            };
        }
        catch
        {
            return null;
        }
    }
}
